feat(consensus): Improve synchronizer

NaitsabesMue opened the pull request:
# Description of change

This PR improves block sync performance when a single/few missing block blocks may suspend many others.

Problem:
Random selection in the periodic synchronizer often takes too long to locate blocks that are held by only a few authorities, leaving large portions of the DAG stalled.

Fix:
Add a knowledge-based fetch path to the periodic synchronizer:

- Leverage extended missing_blocks map from BlockManager, which now tracks for each missing block a set of authorities that are known to have it.
- Prioritize “knowledge-based” peers in the scheduler: when requesting missing blocks, first query exactly those authorities that should already hold them, rather than picking peers at random.
- Issue random‐peer requests alongside the knowledge‐based queries

This change avoids long delays in recovering blocks that live on only a few nodes by directing periodic sync requests to the right peers up front.



## Links to any relevant issues

closes #7370

## How the change has been tested

- X Basic tests (linting, compilation, formatting, unit/integration tests)
- X Patch-specific tests (correctness, functionality coverage)
- X I have added tests that prove my fix is effective or that my feature works
- X I have checked that new and existing unit tests pass locally with my changes

### Release Notes

- [ ] Protocol:
- [ ] Nodes (Validators and Full nodes):
- [ ] Indexer:
- [ ] JSON-RPC:
- [ ] GraphQL:
- [ ] CLI:
- [ ] Rust SDK:
- [ ] REST API:

NaitsabesMue committed:
start reverting subscription count

test(synchronizer): add multi-block fetch scenario

test(synchronizer): ensure no hot fetch triggers in non-hot regime metrics

fixed names of metrics

fix names

fmt

fix test

fix mocking

Update consensus/core/src/synchronizer.rs

Co-authored-by: Piotr Macek <4007944+piotrm50@users.noreply.github.com>

typo

feat(synchronizer): add “hot” fetch request metrics

- Introduce `synchronizer_requested_blocks_by_peer` and `…_by_authority` with label "hot"

refactor(synchronizer): fetch missing ancestors only in hot-fetch regime

- Introduce HOT_FETCH_RATIO_THRESHOLD as global constant
- Avoid querying `get_missing_ancestors_for` unless in hot-fetch regime

refactor(core-thread): return suspended_blocks count instead of full map

- Rename `GetSuspended` RPC to `GetSuspendedCount`
- Update dispatcher and core-thread handler to send/receive `usize` rather than the full `BTreeMap`

refactor(core-thread): reduce payload of GetMissingAncestors

- Replace `GetMissingAncestors` command with `GetMissingAncestorsFor`, scoped to known missing blocks
- Avoid sending full `missing_ancestors` map to synchronizer
- Update dispatcher trait and mock implementations accordingly
- Adjust tests to reflect new method signature

test(synchronizer): add hot-fetch authority‐selection unit test

- Introduce `SyncMockDispatcher` as a test RPC stub
- Add `hot_fetch_with_multiple_authorities` to simulate 10 nodes with 51 suspended blocks across authorities 2,3,4
- Verify hot‐fetch picks the lowest index (2) deterministically under test

Renamed author_to_blocks to authority_to_blocks

Make SuspendedBlock.block internal and expose author() getter for external access.

corrected missing ratio calculation and removed unncessary scope

refactor(synchronizer): drive hot-fetch via CoreThreadDispatcher instead of BlockManager

- Remove direct `BlockManager` RwLock reads in hot-fetch path
- Use `get_suspended_blocks()` and `get_missing_ancestors()` on `core_dispatcher` to snapshot suspension state

Refactor hot fetch logic.

Improve logging for hot fetches: include block list and authority ID; rename variables for clarity.

Fix test build by avoiding .choose() under #[cfg(test)]

Refactor dependents_of and dependent_count for efficiency and clarity; removed filter_map, added explicit invariant check with expect.

synchronizer: optimize hot-fetch by grouping requests per peer

- Previously, hot-fetch logic sent individual fetches per missing block, causing multiple redundant requests per peer.
- Improved efficiency by grouping hot missing blocks per dependent author, then choosing one random peer and issuing a single combined fetch.
- Added an early-exit check (`num_suspended_blocks`) to avoid unnecessary hot-fetch logic.

fixed format

silenced clippy

synchronizer: add proactive hot‐fetch for high‐fanout missing blocks

- Before the normal batching in `fetch_blocks_from_authorities`, scan for any missing block whose `dependent_count` exceeds `HOT_DEPENDENT_THRESHOLD` (50).
- For each “hot” block, call `dependents_of` and pick one random author to send a single‐block fetch request.
- Preserve the existing MAX_PEERS * MAX_BLOCKS_PER_FETCH logic unchanged.

Add dependent_count and dependents_of to BlockManager with test

- Adds helper methods to retrieve number of suspended blocks depending on a missing block, and to collect their author indices.

- Includes unit test verifying correct behavior in empty and single-dependent cases.

polinikita commented:
We need to figure out how to make a consistent change. You are taking a block_manager as a parameter. However, for retrieving missing blocks in periodic synchronizer, we access these blocks through a comman in a core thread because we wanted to avoid concurrent access for this structure. So, if we have a block_manager in synchronizer, then I would remove a command from the core thread for retrieving missing blocks and read them directly.

polinikita commented on consensus/core/src/block_manager.rs:
```diff
@@ -721,6 +722,30 @@ impl BlockManager {
+            .map(|children| {
+                children
+                    .iter()
+                    .filter_map(|child_ref| self.suspended_blocks.get(child_ref))
```
Is this filter_map needed? I was thinking that we only put suspended block refs insides the value for a missing block

NaitsabesMue replied:
Removed the filter map. Coded the logic directly.

polinikita commented on consensus/core/src/synchronizer.rs:
```diff
@@ -1014,7 +1026,58 @@ impl<C: NetworkClient, V: BlockVerifier, D: CoreThreadDispatcher> Synchronizer<C
+                    // Always bind authors as mutable to allow shuffling in non-test environments.
+                    // In test builds, the `mut` is unused, so we silence Clippy.
+                    #[allow(unused_mut)]
+                    let mut authors = bm.dependents_of(missing);
```
The returned set is a vector where same authors can be considered multiple times. i would suggest to change that function to output each authors at most once or represent multiplicities.

NaitsabesMue replied:
Changed to a unique vector. Might add multiples (and sample weight) later if needed.

polinikita commented on consensus/core/src/synchronizer.rs:
```diff
@@ -1014,7 +1026,58 @@ impl<C: NetworkClient, V: BlockVerifier, D: CoreThreadDispatcher> Synchronizer<C
+                    let mut authors = bm.dependents_of(missing);
+
+                    #[cfg(not(test))]
+                    authors.shuffle(&mut ThreadRng::default());
```
Shuffling might be an expensive operation given that we take only one peer from the set

NaitsabesMue replied:
Thanks, changed to a random index.

polinikita commented on consensus/core/src/synchronizer.rs:
```diff
@@ -1014,7 +1026,58 @@ impl<C: NetworkClient, V: BlockVerifier, D: CoreThreadDispatcher> Synchronizer<C
+                .for_each(|missing| {
+                    // Always bind authors as mutable to allow shuffling in non-test environments.
+                    // In test builds, the `mut` is unused, so we silence Clippy.
+                    #[allow(unused_mut)]
```
I think that too many flags inside this function. This makes it harder to understand. Also some unncessary discussion below on reasoning why the flag is needed

polinikita commented on consensus/core/src/synchronizer.rs:
```diff
@@ -1014,7 +1026,58 @@ impl<C: NetworkClient, V: BlockVerifier, D: CoreThreadDispatcher> Synchronizer<C
+                        // is clear, safe, and equivalent. We explicitly
+                        // allow Clippy's warning for clarity and correctness.
+                        #[allow(clippy::clone_on_copy)]
+                        let block_set = std::iter::once(missing.clone()).collect::<BTreeSet<_>>();
```
For every such block we make a separete request. This should work differently and we need to aggregate all requests for the same peer

NaitsabesMue replied:
Changed the logic, choosing a random authority and then collecting all missing blocks it should know.

NaitsabesMue commented:
> We need to figure out how to make a consistent change. You are taking a block_manager as a parameter. However, for retrieving missing blocks in periodic synchronizer, we access these blocks through a comman in a core thread because we wanted to avoid concurrent access for this structure. So, if we have a block_manager in synchronizer, then I would remove a command from the core thread for retrieving missing blocks and read them directly.

 - Removed direct `BlockManager` RwLock reads in hot-fetch path  
- Use `get_suspended_blocks()` and `get_missing_ancestors()` on `core_dispatcher` to snapshot suspension state

piotrm50 commented on consensus/core/src/synchronizer.rs:
```diff
@@ -1014,7 +1029,67 @@ impl<C: NetworkClient, V: BlockVerifier, D: CoreThreadDispatcher> Synchronizer<C
+        {
+            if suspended_blocks.len() > HOT_FETCH_RATIO_THRESHOLD {
+                // Step 1: Map authors to their missing blocks
+                let mut author_to_blocks: HashMap<AuthorityIndex, Vec<BlockRef>> = HashMap::new();
```
Variable name `author_to_blocks` is not entirely correct, as it also contains authorities that reference the missing block. Maybe rename to `authorities_to_blocks`.

NaitsabesMue replied:
changed

piotrm50 commented on consensus/core/src/synchronizer.rs:
```diff
@@ -1014,7 +1029,67 @@ impl<C: NetworkClient, V: BlockVerifier, D: CoreThreadDispatcher> Synchronizer<C
+        // blocks from the authors of blocks referencing them.
+        const HOT_FETCH_RATIO_THRESHOLD: usize = 50;
+        {
+            if suspended_blocks.len() > HOT_FETCH_RATIO_THRESHOLD {
```
Ratio is not calculated, but a length of suspended blocks is used.

NaitsabesMue replied:
Thanks. The ratio calculation got lost.

piotrm50 commented on consensus/core/src/synchronizer.rs:
```diff
@@ -1156,10 +1228,34 @@ impl<C: NetworkClient, V: BlockVerifier, D: CoreThreadDispatcher> Synchronizer<C
+            let sb = suspended_blocks
+                .get(dependent)
+                .expect("Suspended block for missing ancestor should exist.");
+            let author = sb.block.author();
```
Instead of exposing `block` we can create a getter to retrieve `author` directly.

NaitsabesMue replied:
done

piotrm50 commented on consensus/core/src/synchronizer.rs:
```diff
@@ -1014,7 +1029,67 @@ impl<C: NetworkClient, V: BlockVerifier, D: CoreThreadDispatcher> Synchronizer<C
+        // HOT_FETCH_RATIO_THRESHOLD, proactively fetch a batch of missing
+        // blocks from the authors of blocks referencing them.
+        const HOT_FETCH_RATIO_THRESHOLD: usize = 50;
+        {
```
Why the additional scope?

NaitsabesMue replied:
Some artefact. Removed.

piotrm50 commented on consensus/core/src/synchronizer.rs:
```diff
@@ -1014,7 +1029,67 @@ impl<C: NetworkClient, V: BlockVerifier, D: CoreThreadDispatcher> Synchronizer<C
+                let chosen = author_to_blocks.iter().choose(&mut ThreadRng::default());
+
+                #[cfg(test)]
+                let chosen = author_to_blocks.iter().next();
```
Do I understand correctly that we're randomly selecting one authority to fetch a subset of all missing blocks? I think we should send request to the corresponding authorities for all the missing blocks rather than just one.

NaitsabesMue replied:
We randomly select an authority and then request all missing blocks that are referenced by a block authored by this authority.

piotrm50 commented on consensus/core/src/synchronizer.rs:
```diff
@@ -1014,7 +1029,67 @@ impl<C: NetworkClient, V: BlockVerifier, D: CoreThreadDispatcher> Synchronizer<C
+            }
+        }
+
+        // Proceed with the usual fetching
```
Why do we continue with regular fetching if we've already sent requests for all (some?) missing blocks to their authors or authorities that have referenced them?

NaitsabesMue replied:
Not all missing blocks, see above. We could, however, remove these blocks from the subsequent fetches, or increase `MAX_AUTHORITIES_TO_FETCH_PER_BLOCK` by 1.

piotrm50 commented on consensus/core/src/block_manager.rs:
```diff
@@ -683,6 +684,16 @@ impl BlockManager {
 
+    /// Returns all suspended blocks.
+    pub(crate) fn suspended_blocks(&self) -> BTreeMap<BlockRef, SuspendedBlock> {
+        self.suspended_blocks.clone()
```
Another thing we might change is to not send the full SuspendedBlock but rather only its author as a value in the map.

NaitsabesMue replied:
done

piotrm50 commented on consensus/core/src/core_thread.rs:
```diff
@@ -159,6 +176,20 @@ impl CoreThread {
+                            let _scope = monitored_scope("CoreThread::loop::get_missing_ancestor_for");
+                            // Fetch full map from Core, then filter to requested keys.
+                            let full_map = self.core.get_missing_ancestors();
+                            let filtered = keys.into_iter()
```
Maybe we can put this filter inside if that's the only usage of this method.

piotrm50 commented on consensus/core/src/synchronizer.rs:
```diff
@@ -1014,7 +1035,81 @@ impl<C: NetworkClient, V: BlockVerifier, D: CoreThreadDispatcher> Synchronizer<C
+                        let block_hostname =
+                            &context.committee.authority(block_ref.author).hostname;
+                        metrics
+                            .synchronizer_requested_blocks_by_authority
```
We should count normal fetches as well, not only the hot ones with the new metric.

piotrm50 commented on consensus/core/src/synchronizer.rs:
```diff
@@ -1911,4 +2025,189 @@ mod tests {
+        core_disp.stub_missing_ancestors(missing_anc.clone()).await;
+
+        let network = Arc::new(MockNetworkClient::default());
+        // Stub *all* three authorities so none panic:
```
```suggestion
        // Stub *all* ten authorities so none panic:
```

piotrm50 commented on consensus/core/src/synchronizer.rs:
```diff
@@ -1911,4 +2025,189 @@ mod tests {
+        assert_eq!(*hot_peer, AuthorityIndex::new_for_test(2));
+
+        // 8) The periodic fetch went to peer 1
+        let (_periodic_guard, _periodic_bytes, periodic_peer) = &results[1];
```
Why are two blocks returned if there is only one block in `missing_block` set? If it's a parent of the missing block then why is it returned from a different peer?

NaitsabesMue replied:
There are two requests. One via the hot path and another via the periodic path. In the hot path, we choose from those authorities that have created a suspended block referencing the missing blocks. In the periodic, we choose among all authorities.

piotrm50 reviewed (commented):
Let's extend the tests with different cases where we have hot fetches and regular fetched blocks and where we don't have hot fetches at all (where we have both  alot of missing blocks and little missing blocks).

NaitsabesMue committed:
start reverting subscription count

test(synchronizer): add multi-block fetch scenario

test(synchronizer): ensure no hot fetch triggers in non-hot regime metrics

fixed names of metrics

fix names

fmt

fix test

fix mocking

Update consensus/core/src/synchronizer.rs

Co-authored-by: Piotr Macek <4007944+piotrm50@users.noreply.github.com>

typo

feat(synchronizer): add “hot” fetch request metrics

- Introduce `synchronizer_requested_blocks_by_peer` and `…_by_authority` with label "hot"

refactor(synchronizer): fetch missing ancestors only in hot-fetch regime

- Introduce HOT_FETCH_RATIO_THRESHOLD as global constant
- Avoid querying `get_missing_ancestors_for` unless in hot-fetch regime

refactor(core-thread): return suspended_blocks count instead of full map

- Rename `GetSuspended` RPC to `GetSuspendedCount`
- Update dispatcher and core-thread handler to send/receive `usize` rather than the full `BTreeMap`

refactor(core-thread): reduce payload of GetMissingAncestors

- Replace `GetMissingAncestors` command with `GetMissingAncestorsFor`, scoped to known missing blocks
- Avoid sending full `missing_ancestors` map to synchronizer
- Update dispatcher trait and mock implementations accordingly
- Adjust tests to reflect new method signature

test(synchronizer): add hot-fetch authority‐selection unit test

- Introduce `SyncMockDispatcher` as a test RPC stub
- Add `hot_fetch_with_multiple_authorities` to simulate 10 nodes with 51 suspended blocks across authorities 2,3,4
- Verify hot‐fetch picks the lowest index (2) deterministically under test

Renamed author_to_blocks to authority_to_blocks

Make SuspendedBlock.block internal and expose author() getter for external access.

corrected missing ratio calculation and removed unncessary scope

refactor(synchronizer): drive hot-fetch via CoreThreadDispatcher instead of BlockManager

- Remove direct `BlockManager` RwLock reads in hot-fetch path
- Use `get_suspended_blocks()` and `get_missing_ancestors()` on `core_dispatcher` to snapshot suspension state

Refactor hot fetch logic.

Improve logging for hot fetches: include block list and authority ID; rename variables for clarity.

Fix test build by avoiding .choose() under #[cfg(test)]

Refactor dependents_of and dependent_count for efficiency and clarity; removed filter_map, added explicit invariant check with expect.

synchronizer: optimize hot-fetch by grouping requests per peer

- Previously, hot-fetch logic sent individual fetches per missing block, causing multiple redundant requests per peer.
- Improved efficiency by grouping hot missing blocks per dependent author, then choosing one random peer and issuing a single combined fetch.
- Added an early-exit check (`num_suspended_blocks`) to avoid unnecessary hot-fetch logic.

fixed format

silenced clippy

synchronizer: add proactive hot‐fetch for high‐fanout missing blocks

- Before the normal batching in `fetch_blocks_from_authorities`, scan for any missing block whose `dependent_count` exceeds `HOT_DEPENDENT_THRESHOLD` (50).
- For each “hot” block, call `dependents_of` and pick one random author to send a single‐block fetch request.
- Preserve the existing MAX_PEERS * MAX_BLOCKS_PER_FETCH logic unchanged.

Add dependent_count and dependents_of to BlockManager with test

- Adds helper methods to retrieve number of suspended blocks depending on a missing block, and to collect their author indices.

- Includes unit test verifying correct behavior in empty and single-dependent cases.

polinikita committed:
simplify periodic synchronizer

polinikita committed:
fix edge cases

polinikita committed:
add retries in periodic synchronizer

polinikita committed:
metrics

polinikita committed:
fmt

polinikita committed:
fmt

polinikita committed:
fmt

polinikita committed:
fix test and randomization

polinikita committed:
fmt

NaitsabesMue committed:
test(synchronizer): adapted test

polinikita committed:
fix the test and label

NaitsabesMue committed:
metrics(synchronizer): record live‐sync requested blocks per peer and authority

VorobyevIlya commented on consensus/core/src/synchronizer.rs:
```diff
@@ -1018,19 +1052,115 @@ impl<C: NetworkClient, V: BlockVerifier, D: CoreThreadDispatcher> Synchronizer<C
 
+        #[cfg_attr(test, allow(unused_mut))]
+        let mut all_missing_blocks: Vec<BlockRef> = missing_blocks.keys().cloned().collect();
+        #[cfg(not(test))]
+        all_missing_blocks.shuffle(&mut rng);
+
```
Do I understand correctly that from additional random peers we will ask random blocks from missing_block?
From non-random peers, we ask lexicographically first, and it makes sense for me: we want to fetch blocks from round 1 before round 10.
But these non-random guys may also not know blocks from round 1. So for me it makes more sense to ask for a first chunk here too.

polinikita committed:
add new test for missing blocks

polinikita committed:
add new test for missing blocks

polinikita reviewed (approved):
The logic of choosing peers from whom one should fetch the block can be further improved. For instance, one could also take into account from whom we received a block and propogate that knowledge to the block manager. Or the way how blocks are assigned to peers can always follow the rule from smallest-round block to highest round, but now one can leave such improvements for future work.

VorobyevIlya commented on consensus/core/src/block_manager.rs:
```diff
@@ -1108,6 +1151,80 @@ mod tests {
         }
     }
 
+    /// Tests that authorities that know about missing blocks are correctly
```
sentence is not finished

NaitsabesMue replied:
finished it

NaitsabesMue committed:
refine comments and descriptions to match updated fetch logic

NaitsabesMue commented:
> Let's extend the tests with different cases where we have hot fetches and regular fetched blocks and where we don't have hot fetches at all (where we have both alot of missing blocks and little missing blocks).

Tests are extended

NaitsabesMue committed:
Merge branch 'develop' into consensus/feat/improve-synchronizer
